### PR TITLE
feat(search): add overflow menu

### PR DESCRIPTION
**Pull request by cyndichin**
## Summary
User can activate result overflow button

## References 
IN-1123

## Implementation Details
* Create new `OverflowMenu` view for overflow actions (add tags, archive/unarchive and delete). 
* Added `SavedItemsController` so that `SearchViewModel` can observe changes from core data
* Added 5 new search analytic actions

Note: Some issues may still exist and will be done in future tickets
* All Items have both Saves and Archive (archive and deleting is weird for this case) - I would like to discuss with the team on how to handle this and vote a new ticket to be created.
* Resubmitting search after performing an action will cause some inconsistencies in the search result view - ticket for edge case

## Test Steps
- [ ] GIVEN search is active
AND a search has been performed
AND there are 1+ results
AND search results are visible
THEN every search result includes an overflow button

- [ ] GIVEN that a user taps on the search result overflow button 
THEN the search results overflow menu is shown exactly as it would be if the same overflow button were tapped from Saved/Archived

**For All Items scope:**
1. If a user taps on add tags option, the user should be able to see the add tags view.
2. If a user deletes an item, the user should see that the item is deleted and removed the list.
3. If a user archives or moves an item to saves, the user should still see the item in the list, but update the cell to reflect “Archive” or “Move to Saves” depending on the type of item.

## PR Checklist:
- [x] Added Unit / UI tests
- [x] Self Review (review, clean up, documentation, run tests)
- [x] Basic Self QA

## Screenshots
https://user-images.githubusercontent.com/6743397/221930660-3146e072-932c-45c6-92fa-f41ddb7c86ff.mp4

### CODE REVIEW

File: PocketKit/Sources/PocketKit/MyList/SearchItemsList/SearchViewModel.swift [hunk 1]
```diff
@@ -128,15 +129,15 @@ class SearchViewModel: ObservableObject {
...
-            }
+//        userStatusListener = user
+//            .statusPublisher
+//            .receive(on: DispatchQueue.main)
+//            .sink { [weak self] status in
+//                guard self?.searchState?.isEmptyState == true else {
+//                    return
+//                }
+//                self?.searchState = self?.defaultState
+//            }
```
[the remark this answers is not in the document]
Comment (cyndichin) [reply PocketKit/Sources/PocketKit/MyList/SearchItemsList/SearchViewModel.swift:140]: oh this is temporary. 

commit `[fix(tests): changes to fix tests to be able to check if test passes](https://github.com/Pocket/pocket-ios/pull/445/commits/8d21de7b568bced92595cc0d8ee61b8f42727bb8)` is temporary and I'll revert once we fix the issue. However, I wanted to be able to validate the tests that I have currently with Bitrise.

File: PocketKit/Sources/Analytics/AppEvents/Search.swift [hunk 2]
```diff
@@ -93,6 +93,126 @@ public extension Events.Search {
         )
     }
 
+    /**
+     Fired when a user clicks the Delete option in the overflow menu in a card for `Search`
+     */
+    static func deleteItem(
```
Comment (bassrock) [inline PocketKit/Sources/Analytics/AppEvents/Search.swift:99]: so many explicit analytics 🔥 i love it

File: PocketKit/Sources/PocketKit/MyList/ItemsList/PocketItemViewModel.swift [hunk 3]
```diff
@@ -7,78 +7,89 @@ import Analytics
...
-    init(item: PocketItem, index: Int, source: Source, tracker: Tracker) {
+    /// Determines whether an item should show Archive or Move To Saves
+    var isArchived: Bool {
+        item.isArchived
+    }
+
+    /// Retrieves view model to present Add Tags view
+    var tagsViewModel: PocketAddTagsViewModel? {
+        guard let savedItem = fetchSavedItem() else {
+            Log.capture(message: "PocketAddTagsViewModel not returned")
```
Comment (bassrock) [inline PocketKit/Sources/PocketKit/MyList/ItemsList/PocketItemViewModel.swift:33]: 🥳  ty!!!

File: PocketKit/Sources/PocketKit/MyList/ItemsList/PocketItemViewModel.swift [hunk 4]
```diff
@@ -7,78 +7,89 @@ import Analytics
...
         self.isFavorite = item.isFavorite
     }
 
-    func favoriteAction(index: Int, scope: SearchScope) -> ItemAction {
+    /// Triggers action to favorite or unfavorite an item in a list
+    func favoriteAction() -> ItemAction {
         if isFavorite {
             return .unfavorite { [weak self] _ in
-                self?._unfavorite(index: index, scope: scope)
+                guard let self else { return }
```
Comment (bassrock) [inline PocketKit/Sources/PocketKit/MyList/ItemsList/PocketItemViewModel.swift:59]: @cyndichin should we log when self is nil especially since it got you and I the other day.

File: PocketKit/Sources/PocketKit/MyList/ItemsList/SwiftUI/Components/OverflowMenu.swift [hunk 5]
```diff
@@ -2,37 +2,97 @@ import SwiftUI
 import Textile
 
 struct OverflowMenu: View {
-    var overflowActions: [ItemAction]
-    var trackOverflow: ItemAction?
-    var image: ImageAsset = .overflow
-    var trailingPadding: Bool = false
+    @EnvironmentObject
```
Comment (bassrock) [inline PocketKit/Sources/PocketKit/MyList/ItemsList/SwiftUI/Components/OverflowMenu.swift:5]: In the future, lets see how we can do this in a way that enables SwiftPreviews.

File: PocketKit/Sources/PocketKit/MyList/SearchItemsList/SearchViewModel.swift [hunk 6]
```diff
@@ -390,12 +435,67 @@ extension SearchViewModel {
...
+    /// - Parameters:
+    ///   - savedItem: savedItem that was changed in Core Data
+    ///   - items: list of `PocketItem` that is displayed as search results
+    private func updateItemInView(_ savedItem: SavedItem, and items: [PocketItem], at index: Int) {
+        var items = items
+        items.remove(at: index)
+        items.insert(PocketItem(item: savedItem), at: index)
+        Log.debug("Search item updated \(String(describing: savedItem.title))")
+        // Animations seen to work better when we don't wrap this around main thread
+        self.searchState = .searchResults(items)
```
[the remark this answers is not in the document]
Comment (bassrock) [reply PocketKit/Sources/PocketKit/MyList/SearchItemsList/SearchViewModel.swift:489]: I'd love to see this live, not putting UI on main thread scares me.

**Review by bassrock — APPROVED** [review]
@cyndichin I see we added a bunch of analytics but no UI tests for them.

Can we add those in, in this PR or a followup? I don't want to block on it though.

File: PocketKit/Sources/PocketKit/MyList/ItemsList/PocketItemViewModel.swift (hunk 4 above)
Comment (cyndichin) [reply PocketKit/Sources/PocketKit/MyList/ItemsList/PocketItemViewModel.swift:59]: will update!

**Comment by cyndichin** [comment]
> @cyndichin I see we added a bunch of analytics but no UI tests for them.
> 
> Can we add those in, in this PR or a followup? I don't want to block on it though.

@bassrock I added the UI tests with analytics attached, should be in `SearchTests`. The file was a large diff, so Github may have hidden it. Let me know if you still don't see it. Thanks!

**Comment by bassrock** [comment]
@cyndichin you are right! Thank you!

File: PocketKit/Sources/PocketKit/MyList/ItemsList/SwiftUI/Components/OverflowMenu.swift (hunk 5 above)
Comment (cyndichin) [reply PocketKit/Sources/PocketKit/MyList/ItemsList/SwiftUI/Components/OverflowMenu.swift:5]: yes, should I create a ticket for this or keep in mind when creating SwiftUI views in the future?
